### ISSUE TITLE
User synchronisation without LDAP attributes

### ISSUE THREAD

**Issue by surfi2000** [piwik/plugin-LoginLdap#132]
Would like ability to synchronise users without the need to customise LDAP attributes.

On admin side would like to map an LDAP group to a website and the access level to that website

**Comment by nebevservian** [comment]
I agree. It would be much more standard to have entitlements based on groups, rather than on random LDAP attributes. My motivation for *not* using attributes include:

- This often involves modifying the schema
- Once the schema is modified, you cannot go back (in Active Directory anyway)
- Something needs to programatically generate the permission strings
    - This becomes especially difficult when dealing with multiple installations
    - You would have to be aware of all instances in your organisation in order for this to work
- Especially in large organisations, the chances of modifying the core schema are essentially 0

Groups are a standard. Everyone knows how to use them, and most other applications use them for entitlements. I don't see why Piwik should be any different. Rather than merely complain about it, I made a fork.

https://github.com/nebevservian/plugin-LoginLdap/tree/feature/group-entitlements

This fork allows you to:
- Specify a _superuser_ DN - All users part of this group will be superusers
- Specify _admin_ and _view_ DNs for each site that is part of the installation

If you use recursive groups (groups that a member of other groups) and you _don't_ use Active Directory, this won't work. If you do, I think this would solve your problem.

**Comment by mattab** [comment]
Hi @surfi2000 @mikechamberlain maybe you could test the proposal of @nebevservian in his fork at https://github.com/nebevservian/plugin-LoginLdap/tree/feature/group-entitlements ?

@nebevservian great to see your contribution! hopefully other users will test and report feedback, and maybe it could be merged later

**Comment by mikechamberlain** [comment]
Hi guys,

Happy to be contributing to the project. 

As well as adding the LDAP groups stuff we took the opportunity to restructure the plugin HTML to align it to the modern Piwik styleguide. 

I'm guessing that the whole thing may need a little more work to bring it up to release quality. For instance, the docs need updating and there are no unit tests for the new code yet.

**Comment by corecache** [comment]
What happened to these ambitions?

**Comment by surfi2000** [comment]
@mikechamberlain could you share what you have? 

I actually ended up writing a little ruby script that syncs access between LDAP and the Piwik DB directly to work around this issue but would much prefer to have entitlements based on groups built into the plugin. 

@mattab I'll take a look at @nebevservian fork and report back. It sounds like it has all the requirements

**Comment by mikechamberlain** [comment]
Hi @surfi2000, I work with @nebevservian so all my work is part of https://github.com/nebevservian/plugin-LoginLdap. The backend has been updated to work with Piwik 3, but I think the front end will need some attention to make it look right. We script our Piwik LDAP configuration now so it's not likely we'll be prioritizing this any time soon.

**Comment by mattab** [comment]
If anyone wants to contribute this feature we will gladly accept & review pull request :+1:

**Comment by thungsten** [comment]
Hi!
What is the status? Can the forked version be used?
Would be a great help at least to separate admin group from average users group...
Even better if recursive/nested groups in active directory work as well!

**Comment by nebev** [comment]
We're using the plugin in production. It does support nested AD groups too.
However, I'm not sure it's going to meet the community standards in terms of quality. We also had to rewrite a lot of the UI components to make in order for it to work, so there's likely to be a big diff.

**Comment by thungsten** [comment]
How did you configure it to accept nested groups - we recently tried to use nested groups (two groups of departments have been added to the piwik user group) but the nested groups weren't authorised.
So we still add each member to a generic piwik user group. After the initial login of an user the user is manually assigned to a proper set of groups by one of the admins.

Or do you talk specifically about the fork?
I downloaded it and installed it but it ended up with piwik not being accessible for leap users (all the time error screens). Did I miss something?

Indeed, we would be glad to be able to assign certain groups to reported sites - that way it wouldn't chew necessary to login before I/an admin can assign access rights...

**Comment by thungsten** [comment]
@nebevservian can you provide any further information what has to done?

**Comment by nebev** [comment]
Hey mate. On the `group-entitlements-v3` branch of the fork, have a look at https://github.com/nebevservian/plugin-LoginLdap/blob/feature/group-entitlements-v3/LdapInterop/UserAccessGroupMapper.php
Specifically the method `userExistsRecursive`. This is an Active Directory feature that allows you to determine if a user is a member of a group via membership of another group. It works for us, but we're using an older version of Piwik (3.1.1).

**Comment by dougti** [comment]
Just adding my support for these features to be included in the official plugin.

Group based auth with support for nested groups is how we manage all of our AAA here.

**Comment by petr-zeman** [comment]
I also vote for adding the group based rules into the official LdapLogin.

**Comment by cweiland** [comment]
A little UP to have this feature on official LdapLogin, and same thing with SAML one (even if it not the right place)
